src/fairseq2/cluster.py

```diff
@@ -182,3 +186,163 @@ def supports_current_cluster(self) -> bool:
     @override
     def supported_cluster(self) -> str:
         return "none"
+
+
+@ray.remote(num_cpus=0)
+class RayCoordinator:
+    NAME = "RAY_FAIRSEQ2_COORDINATOR_NAME"
+    LEADER_MAX_RETRIES = 30
+    LEADER_RETRY_INTERVAL = 1.0
+
+    def __init__(self, job_id, num_nodes, gpus_per_node, placement_group_ids):
+        self.job_id = job_id
+        self.worker_info = {}
+        self.num_nodes = num_nodes
+        self.gpus_per_node = gpus_per_node
+        self.leader = None
+        self.leader_port = None
+        self.ready_workers = 0
+        self.total_workers = num_nodes * gpus_per_node
+        self.placement_group_ids = placement_group_ids
+
+    def register_worker(self, worker_id, hostname, node_id, placement_group_id, gpu_id):
+        """Register a worker with its placement group ID and GPU ID"""
+        assert (
+            placement_group_id in self.placement_group_ids
+        ), f"Placement group id {placement_group_id} should in {self.placement_group_ids}"
+        assert 0 <= gpu_id <= 7, f"Bad gpu_id {gpu_id}"
+        node_index = self.placement_group_ids.index(placement_group_id)
+        global_rank = node_index * self.gpus_per_node + gpu_id
```

Comment:
would it be working if ray cluster does not take a full node ?

Reply:
this version is not going to work as it assumes gpu 0 is the leader. Just updated the code to passing in the rank, where rank==0 is the leader. So this will enable we asking for not all the gpus of a node.